proxy/src/control/observe.rs

```diff
@@ -156,3 +163,24 @@ impl Drop for TapEvents {
         }
     }
 }
+
+impl Eq for RequestById {}
+
+impl PartialEq for RequestById {
+    fn eq(&self, other: &RequestById) -> bool {
+        self.req.id.eq(&other.req.id)
+    }
+}
+
+impl Hash for RequestById {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.req.id.hash(state)
+    }
+}
+
+impl Equivalent<RequestById> for ctx::http::RequestId {
```

Comment:
Note that I think we can/should eventually abstract this pattern out into a reusable trait, something like `trait IdentifiedByField<T: Hash + Eq + PartialEq + Clone> { id(&'a self) -> 'a T }`, but I didn't want to bog this down.